main.c

```diff
@@ -167,13 +167,29 @@ static bool match_refresh(const struct kanshi_mode *mode, int refresh) {
 static struct kanshi_mode *match_mode(struct kanshi_head *head,
 		int width, int height, int refresh) {
 	struct kanshi_mode *mode;
+	struct kanshi_mode *last_match = NULL;
+
 	wl_list_for_each(mode, &head->modes, link) {
-		if (mode->width == width && mode->height == height &&
-				(refresh == 0 || match_refresh(mode, refresh))) {
-			return mode;
+		if (mode->width == width && mode->height == height) {
+			if(refresh && match_refresh(mode, refresh)) {
+				return mode;
+			}
+
+			if (refresh == 0) {
+				if(!last_match) {
```

Comment:
This can be simplified to:

```c
if (!last_match || mode->refresh > last_match->refresh) {
```

This would also make the `continue` statement unnecessary.